### PR TITLE
Firehose destination

### PR DESCRIPTION
**Pull request by ruchiramoitra**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [x] I have made corresponding changes to the documentation

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-server/412)

### CODE REVIEW

File: services/streammanager/firehose/firehosemanager.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+			putOutput, errorRec = fh.PutRecord(&firehose.PutRecordInput{
+				DeliveryStreamName: aws.String(deliveryStreamMap[i]["to"]),
+				Record:             &firehose.Record{Data: value},
+			})
+		}
+		if errorRec != nil {
+			if awsErr, ok := errorRec.(awserr.Error); ok {
+				if reqErr, ok := errorRec.(awserr.RequestFailure); ok {
+					logger.Errorf("error in firehose :: %v", awsErr.Code())
+					fmt.Println(reqErr.StatusCode(), reqErr.RequestID())
```
[the remark this answers is not in the document]
Comment (ruchiramoitra) [reply services/streammanager/firehose/firehosemanager.go:89]: done

File: services/streammanager/firehose/firehosemanager.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		return statusCode, err.Error(), err.Error()
+	}
+	putOutput = nil
+	for i := 0; i < len(deliveryStreamMap); i++ {
+		var statusCode int
+		if event.Value() == deliveryStreamMap[i]["from"] {
+			putOutput, errorRec = fh.PutRecord(&firehose.PutRecordInput{
+				DeliveryStreamName: aws.String(deliveryStreamMap[i]["to"]),
+				Record:             &firehose.Record{Data: value},
+			})
```
[the remark this answers is not in the document]
Comment (ruchiramoitra) [reply services/streammanager/firehose/firehosemanager.go:83]: done

File: services/streammanager/firehose/firehosemanager.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	AccessKeyID string
+	AccessKey   string
+	MapEvents   []map[string]string
+}
+
+var putOutput *firehose.PutRecordOutput = nil
+var errorRec error
+var event, typeCall gjson.Result
+
+func init() {
```
[the remark this answers is not in the document]
Comment (ruchiramoitra) [reply services/streammanager/firehose/firehosemanager.go:28]: done

File: config/sample.env [hunk 4]
```diff
@@ -88,4 +88,4 @@ WAREHOUSE_URL=http://localhost:8082
 # The below keys are used to give access to upload, download from s3 bucket and also used by copy command for REDSHIFT
 # Use this if you don't to not give aws cred in control plane ex: S3, REDSHIFT
 # RUDDER_AWS_S3_COPY_USER_ACCESS_KEY_ID=<rudder user access key>
-# RUDDER_AWS_S3_COPY_USER_ACCESS_KEY=<rudder user access key secret>
+# RUDDER_AWS_S3_COPY_USER_ACCESS_KEY=<rudder user access key secret>
```
[the remark this answers is not in the document]
Comment (ruchiramoitra) [reply config/sample.env:91]: done

File: rudder-docker.yml [hunk 5]
```diff
@@ -9,35 +8,35 @@ services:
       - POSTGRES_DB=jobsdb
     ports:
       - "6432:5432"
-  backend:
```
[the remark this answers is not in the document]
Comment (ruchiramoitra) [reply rudder-docker.yml:12]: done

File: services/streammanager/firehose/firehosemanager.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+			statusCode := 500
+			return statusCode, err.Error(), err.Error()
+		}
+
+		deliveryStreamMapToInputString := strings.Trim(string(deliveryStreamMapToInput), "\"")
+
+		putOutput, errorRec = fh.PutRecord(&firehose.PutRecordInput{
+			DeliveryStreamName: aws.String(string(deliveryStreamMapToInputString)),
+			Record:             &firehose.Record{Data: value},
+		})
```
Comment (prabrishac) [inline services/streammanager/firehose/firehosemanager.go:84]: Rather use this - 
deliveryStreamMapToInputString, ok := parsedJSON.Get("deliveryStreamMapTo").Value().(string)
if !ok {
		// handle error
	}

File: services/streammanager/firehose/firehosemanager.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	if parsedJSON.Get("deliveryStreamMapTo").Value() != nil {
+		deliveryStreamMapTo := parsedJSON.Get("deliveryStreamMapTo").Value().(interface{})
+		deliveryStreamMapToInput, err := json.Marshal(deliveryStreamMapTo)
+		if err != nil {
+			logger.Errorf("error in firehose :: %v", err.Error())
+			statusCode := 500
+			return statusCode, err.Error(), err.Error()
+		}
+
+		deliveryStreamMapToInputString := strings.Trim(string(deliveryStreamMapToInput), "\"")
```
Comment (prabrishac) [inline services/streammanager/firehose/firehosemanager.go:79]: Rather use this -
deliveryStreamMapToInputString, ok := parsedJSON.Get("deliveryStreamMapTo").Value().(string)
if !ok {
// handle error
}

File: services/streammanager/streammanager.go [hunk 8]
```diff
@@ -21,6 +22,9 @@ func NewProducer(destinationConfig interface{}, destination string) (interface{}
 	case "AZURE_EVENT_HUB":
 		producer, err := kafka.NewProducerForAzureEventHub(destinationConfig)
 		return producer, err
+	case "FIREHOSE":
+		producer, err := firehose.NewProducer(destinationConfig)
+		return producer, err
 	default:
```
Comment (arnab-p) [inline services/streammanager/streammanager.go:28]: As we are adding more and more these things, might be a better idea to keep these sorted in alphabetical order? Will be easier to get to the section when needed.

File: vendor/modules.txt [hunk 9]
```diff
@@ -83,10 +84,13 @@ github.com/bugsnag/bugsnag-go/headers
 github.com/bugsnag/bugsnag-go/sessions
 # github.com/bugsnag/panicwrap v1.2.0
 github.com/bugsnag/panicwrap
+<<<<<<< HEAD
```
Comment (arnab-p) [inline vendor/modules.txt:87]: Rebase issue?

File: services/streammanager/firehose/firehosemanager.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	if parsedJSON.Get("deliveryStreamMapTo").Value() != nil {
+		deliveryStreamMapToInputString, ok := parsedJSON.Get("deliveryStreamMapTo").Value().(string)
+		if !ok {
+			logger.Errorf("error in firehose :: %v", err.Error())
+			statusCode := 500
+			return statusCode, err.Error(), err.Error()
+		}
+		logger.Infof(deliveryStreamMapToInputString)
```
Comment (prabrishac) [inline services/streammanager/firehose/firehosemanager.go:76]: It is only for debugging, right? lets remove it.

File: go.mod [hunk 11]
```diff
@@ -17,8 +17,7 @@ require (
 	github.com/bitly/go-simplejson v0.5.0 // indirect
```
Comment (arnab-p) [inline go.mod:17]: Are we sure about the impact of these changes? @ruchiramoitra @prabrishac

**Pull request merged**